src/cp/cp.rs

```diff
@@ -745,14 +851,68 @@ fn copy_directory(root: &Path, target: &Target, options: &Options) -> CopyResult
         if path.is_dir() && !local_to_target.exists() {
             or_continue!(fs::create_dir_all(local_to_target.clone()));
         } else if !path.is_dir() {
-            copy_file(path.as_path(), local_to_target.as_path(), options)?;
+            if preserve_hard_links {
+                #[cfg(unix)]
+                unsafe {
```

Comment:
Both this `#[cfg(unix)]` section and the `#[cfg(windows)]` section below seem to share code with the sections in `copy()`.  Maybe try to move some of the code into common functions?